src/osd/PG.h

```diff
@@ -2395,10 +2403,75 @@ class PG : DoutPrefixProvider {
   virtual void agent_delay() = 0;
   virtual void agent_clear() = 0;
   virtual void agent_choose_mode_restart() = 0;
+
+  virtual bool can_handle_fast(OpRequestRef& op) = 0;
 };
 
 ostream& operator<<(ostream& out, const PG& pg);
 
 ostream& operator<<(ostream& out, const PG::BackfillInterval& bi);
 
+// for no pg lock completion
+enum COMP_ACK_TYPE {
+  OP_COMP_ACK,
+  OP_COMP_ACK_COMP,
+};
+enum COMP_OP_TYPE {
+  OP_COMP_PRIMARY_OP,
+  OP_COMP_REP_OP,
+  OP_COMP_NO_OP,
+};
+
+class CompletionItem {
+public:
+  /* completion lock */
+  Mutex comp_lock;
+
+// need PG lock
+  /* update_peer_last_complete_ondisk() from REPOP_REPLY */
+  pg_shard_t fromosd;
+  eversion_t lcod;
+  
+  /* last_update_applied */
+  eversion_t applied_version;
+
+  /* callback func which need pg lock */
+  list<Context*> on_applied_with_pg_lock;
+
+  /* Replica versions in MSG_OSD_REPOP */
+  eversion_t rep_committed_version;
+  eversion_t rep_applied_version;
+  int ackerosd;
+
+  /* From OnOpCommit, OnOpApply */
+  pg_shard_t shard;
+  version_t epoch;
+
+  /* RepGather */
+  void * rep_gather;
+
+// don't need PG lock
+  ceph_tid_t rep_tid;
+  int comp_type; // 1: primary 2: replica
+  OpRequestRef op;
+
+  bool all_committed;
+  bool all_applied;
+  bool completed;
+  bool do_reserve_check_completion;
+  utime_t now;
+
+  CompletionItem(int comp_type): comp_lock("CompItem::comp_lock"), rep_gather(NULL), comp_type(comp_type) {
+    all_committed = false;
+    all_applied = false;
+    completed = false;
+    do_reserve_check_completion = false;
+    ackerosd = 0;
+  }
+
+  void register_on_applied_with_pg_lock(Context *c) {
+    if (!c) return;
+    on_applied_with_pg_lock.push_back(c);
+  }
```

Comment:
Well, RepModify is the replica side, right?  Why would that require all_committed or all_applied?  I'm not sure why you want a queue?  The replica write is handled 3 times: once by the original message and twice by the two objectstore callbacks (which can simply reference the RepModify as they currently do).  None of those seem to require an extra queue or an extra data structure.